logr.go

```diff
@@ -239,7 +239,53 @@ func (l Logger) skipSink() bool {
 func (l Logger) skipContextValues() bool {
 	// WithContextValues ensures that l.contextReaders is never set
 	// to an empty slice.
-	return l.ctx == nil || l.contextReaders == nil
+	return l.context == nil || l.context.ctx == nil || l.context.readers == nil
+}
+
+// setContext stores the context and updates any related fields. It mutates the
+// logger and thus is only safe to use for loggers that are not currently being
+// used concurrently.
+func (l *Logger) setContext(ctx context.Context) {
+	if l.context == nil {
+		l.context = &contextMeta{ctx: ctx}
+	} else {
+		cp := &contextMeta{ctx: ctx, readers: l.context.readers}
```

Comment:
Indeed we don't need to copy because `mergeReaders` will copy `l.context.readers` instead of appending to it, so there's no conflict with other loggers.

If there is one feature that I hate about Go, then it's slices. Way too many footguns associated with them.

Reply:
Done.